Adding DID Document Builder

andorsk opened the pull request:
In reference to https://github.com/TBD54566975/ssi-sdk/issues/283

andorsk commented:
@decentralgabe can you take a look and merge if this is ok?

andorsk commented on did/builder.go:
```diff
@@ -0,0 +1,157 @@
+		return errors.Wrap(err, "malformed context")
+	}
+	uniqueContexts := util.MergeUniqueValues(builder.contexts, res)
+	builder.contexts = uniqueContexts
```
great question. I used https://github.com/TBD54566975/ssi-sdk/blob/main/credential/builder.go as a reference. The reason it was suggested was the following: 

On : https://github.com/TBD54566975/ssi-sdk/blob/main/credential/builder.go#L25 	
// contexts and types are kept to avoid having cast to/from interface{} values

I made an assumption here since the context is also the same data type, this is a reasonable implementation to rely upon for did document builder.

decentralgabe commented on did/builder_test.go:
```diff
@@ -0,0 +1,138 @@
+// Exercise all builder methods
+func TestDIDDocumentBuilder(t *testing.T) {
+	// https://www.w3.org/TR/did-core/#example-did-document-with-1-verification-method-type
+	var exampleAuthenticationEntry = map[string]string{
```
please switch these var declarations to use `:=` syntax

decentralgabe replied:
or add a single `var ( ) ` block

andorsk replied:
sounds good.

decentralgabe commented on did/builder_test.go:
```diff
@@ -0,0 +1,138 @@
+		"publicKeyMultibase": "z9hFgmPVfmBZwRvFEyniQDBkz9LmV7gDEqytWyGZLmDXE",
+	}
+
+	builderBad := DIDDocumentBuilder{}
```
```suggestion
	var builderBad DIDDocumentBuilder
```

andorsk replied:
nice callout!

decentralgabe reviewed (approved):
supportive of this!

andorsk commented:
Thanks for the review @decentralgabe . addressed the comments. Realized you might be busy. No rush here. @nitro-neal lmk if your concerns are addressed as well. 

Thanks guys!

andorsk commented:
bonk. @decentralgabe or @nitro-neal can you get back and possible merge if you're good with this?

andresuribe87 commented on did/builder_test.go:
```diff
@@ -0,0 +1,138 @@
+)
+
+// Exercise all builder methods
+func TestDIDDocumentBuilder(t *testing.T) {
```
Would it make sense to split this test into multiple smaller ones, that exercise a single behavior?

andorsk replied:
Reasonable suggestion if we want to update that. I can put a note in for a future PR to move this into smaller tests.

andresuribe87 replied:
Why not do it in this PR?

From a maintainability perspective, it's easier to have smaller, self contained tests. 

FWIW, I want to add coding guidelines into the contribution doc. See https://github.com/TBD54566975/ssi-sdk/issues/290

andorsk replied:
@andresuribe87 again, this was all using the reference of the VC Builder code, which does tests in a similar way. https://github.com/TBD54566975/ssi-sdk/blob/main/credential/builder_test.go#L254. I was being consistent with the existing codebase. I think if we want to break things out, we should also address the VC Builder as well.

andresuribe87 reviewed (commented):
I understand that this is following the same pattern that existed, but it is strange to me that every setter method returns an error, which can only be the BuilderEmptyError. I think it would be cleaner for only `Build` to return an error. The rest should simply accumulate changes.

2. The DIDDocument field is can be easily modified by clients, which

andorsk commented on did/builder.go:
```diff
@@ -0,0 +1,158 @@
+type DIDDocumentBuilder struct {
+	contexts []string
+	types    []string
+	*DIDDocument
```
@andresuribe87 I was following precedence from previous code implemented in the ssi-sdk: https://github.com/TBD54566975/ssi-sdk/blob/main/credential/builder.go#L24.

andorsk commented on did/builder.go:
```diff
@@ -0,0 +1,158 @@
+	return reflect.DeepEqual(builder, &DIDDocumentBuilder{})
+}
+
+func (builder *DIDDocumentBuilder) AddContext(context interface{}) error {
```
@andresuribe87 as long as this works, I'm fine moving to any, but note https://github.com/TBD54566975/ssi-sdk/blob/main/credential/builder.go#L68 uses the `interface{}` type.

andresuribe87 replied:
Either works, filed https://github.com/TBD54566975/ssi-sdk/issues/289

andorsk commented on did/builder.go:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	if err := builder.DIDDocument.IsValid(); err != nil {
+		return nil, errors.Wrap(err, "did doc not ready to be built")
```
sgtm. That works IMO.

andorsk commented on did/builder.go:
```diff
@@ -0,0 +1,158 @@
+// Builds the DID Document
+func (builder *DIDDocumentBuilder) Build() (*DIDDocument, error) {
+	if builder.IsEmpty() {
+		return nil, errors.New(BuilderEmptyError)
```
@andresuribe87 understand. Following existing implementation here https://github.com/TBD54566975/ssi-sdk/blob/main/credential/builder.go#L70 to stay consistent. If we want to revisit how errors are handle in the builders I'm OK with that, but I think they should be consistent across the VC builder and the DIDDocument builder.

decentralgabe commented:
thanks @andorsk !

andorsk commented:
Thanks @andresuribe87 @nitro-neal @decentralgabe for the review!
